drivers: video: Himax HM01B0 camera sensor driver

NinoSc opened the pull request:
Add Himax HM01B0 camera sensor driver.
It depends on I2C and it is required to configure the camera.
reference PR: https://github.com/zephyrproject-rtos/zephyr/pull/93688

NinoSc committed:
drivers: video: Himax HM01B0 camera sensor driver

Add Himax HM01B0 camera sensor driver.
It depends on I2C and it is required to configure the camera.

Signed-off-by: Antonino Scarpaci <antonino.scarpaci@gmail.com>

NinoSc commented:
> Thank you for splitting it out of #93688!
> 
> I think the last thing it needs is an entry in the test framework so that this driver is checked automatically for build errors.
> 
> You can insert a devicetree entry for it in this file:
> 
> * https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/video/app.overlay
> 
> Then this should be enough for the CI to build the driver on every pull request.

I hope it is ok https://github.com/NinoSc/zephyr/blob/42a6083cf50e3e4580ed8d265084a3a896667cc3/tests/drivers/build_all/video/app.overlay#L136
Let me know please

KurtE commented:
Quick questions, observations:   As mentioned in #93688 

Wondering if you have it running on any specific hardware and have it displaying or saving any frames?  
Other than running a quick test with 1 bit data to see how it was differed in the data generated (using Saleae Logic Analyzer)
I have generated valid frames for 4 bit and 8 bit mode, so far I have only tried 320x240

Example on GIGA 4 bit mode (Arducam Camera for GIGA) There are 244 rows of data generated.
<img width="892" height="736" alt="image" src="https://github.com/user-attachments/assets/23b9b267-064c-4567-85c0-154c4d1977ad" />

Each row generates 652 bytes of data.  Two bytes per pixel so 356 pixels per row:
<img width="644" height="747" alt="image" src="https://github.com/user-attachments/assets/ab0bf8d9-e6cd-4acb-950a-24e2bcee6544" />

For the heck of it, I just now tried 160x120 and:
<img width="624" height="403" alt="image" src="https://github.com/user-attachments/assets/f710d442-a4bd-4594-a90c-24abc35cf0a5" />
Looks like it generated 122 columns
<img width="507" height="442" alt="image" src="https://github.com/user-attachments/assets/b0da972d-b8f6-420f-9c2a-39672a7705d0" />
By maybe 328 clocks or 164 pixels... Will try that out...
Added that resolution using the 160x120 init...  and verified that it works on the giga 4 bit mode.
I scaled the pixels up when I displayed them.
<img width="573" height="736" alt="image" src="https://github.com/user-attachments/assets/fb99451d-ca94-4e04-8bf8-adaf3c3afe20" />

But all of this was on STM32H7... processor And working around issues of 4 bits of data per clock / DMA transfer and how do you
combine the two nibbles that come in separate bytes back into one byte.  Some STM32 specific issues/ideas I tried to
mention in a new issue: #94907

However the number for rows/columns generated per the settings I would assume is the same for most processors.

Will be fun to experiment once this is merged...

nixward commented on drivers/video/hm01b0.c:
```diff
@@ -0,0 +1,308 @@
+#define HM01B0_INIT(inst)                                                                          \
+	const struct hm01b0_config hm01b0_config_##inst = {                                        \
+		.i2c = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.data_bits = DT_INST_PROP(inst, data_bits),                                        \
```
Instead of the runtime if statement you could shrink the driver flash usage a small amount with:
```suggestion
		.ctrl_val = HM01B0_CTRL_VAL(DT_INST_PROP(inst, data_bits)),      \
```

And:

```
#define HM01B0_CTRL_VAL(data_bits) \
    ((data_bits) == 8 ? 0x02 : \
     (data_bits) == 4 ? 0x42 : \
     (data_bits) == 1 ? 0x22 : 0x00)
```

The Devicetree enum will ensure only 1, 4 or 8 is valid for the `data-bits` property.

KurtE commented:
> Hello @KurtE. I did not [find `border_px` anywhere](https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20border_px&type=code), could you point at the source that uses it? Is it from this current PR code? Thanks.

Sorry, it is not actually in this PR, but in the PR that this was split from, where he processes the data from this driver.
https://github.com/NinoSc/zephyr/blob/add_hm01b0_sensor_and_video_pio_dma/drivers/video/video_pio_dma.c#L506

In my hacked up version of this PR, I added more formats, like:
```
HM01B0_VIDEO_FORMAT_CAP(164, 122, VIDEO_PIX_FMT_GREY),
HM01B0_VIDEO_FORMAT_CAP(326, 244, VIDEO_PIX_FMT_GREY),

```
https://github.com/KurtE/zephyr/blob/hm01b0/drivers/video/hm01b0.c#L353-L366

Which were the actual sizes of the images returned when you asked for 160x120 and 320x240, at least when I tried them
out on the video_stm32_dcmi.c driver.  Not sure if that is the right answer for this one.

I also added other formats, for when the driver is set into 4 bit mode, as to trick the system to know that 16 bits
are returned for each pixel and not 8... (I did not deal at all with the 1 bit serial mode), although did verify by
logic analyzer that it would return 64 bits (8 pixel clocks) per pixel.

josuah commented:
Thank you! It seems like the Github search is broken as it's still there on the [PIO driver PR](https://github.com/zephyrproject-rtos/zephyr/pull/93688/files#diff-99f40a82855620c62893ebfc0fb59d8c48cf025f165cae92b9a46b920711c5f4R101-R102).

It might be needed to update the image sensor size declared to be what it really is, and remove the cropping from the PIO PR, but that is for #93688. I will follow-up on both PR, thanks!

josuah reviewed (dismissed):
IMHO despite some possible small adjustment, this is ready for being merged.

The details left to be fixed might eventually be addressed as bugfix PRs, unless @NinoSc you wanted to address some before this is merged?

NinoSc commented on dts/bindings/video/himax,hm01b0.yaml:
```diff
@@ -0,0 +1,39 @@
+properties:
+  data-bits:
+    type: int
+    default: 0x1
```
Added default value documentation. Thanks!

KurtE commented:
> It is usually possible with some 2D DMA engines to configure a stride to manage this scenario. The PIO DMA engine actually is configured to manage the boarders.

Assuming that is true, my questions would be, how would my driver know it needs to do so? This also applies to your driver.
That is with the PICO what happens if I use a different camera like:
https://www.amazon.com/Arducam-Camera-Raspberry-HM0360-Module/dp/B098QRKGKB
or maybe (although not sure if it would work same way?)
https://www.amazon.com/Arducam-Module-Megapixels-Arduino-Mega2560/dp/B012UXNDOY

Should your DMA driver strip two pixels as being a border?

NinoSc commented:
> > It is usually possible with some 2D DMA engines to configure a stride to manage this scenario. The PIO DMA engine actually is configured to manage the boarders.
> 
> Assuming that is true, my questions would be, how would my driver know it needs to do so? This also applies to your driver. That is with the PICO what happens if I use a different camera like: https://www.amazon.com/Arducam-Camera-Raspberry-HM0360-Module/dp/B098QRKGKB or maybe (although not sure if it would work same way?) https://www.amazon.com/Arducam-Module-Megapixels-Arduino-Mega2560/dp/B012UXNDOY
> 
> Should your DMA driver strip two pixels as being a border?

I don't have the solution, maybe we can add some properties to the device tree of the dma engine to manage horizontal and vertical borders. My point was that I have seen many FPGA IPs where the stride was a configurable parameter useful to manage this scenario.

josuah commented:
> or maybe (although not sure if it would work same way?)

The OV2640 MCU has an intermediate chip taking the parallel input and pushing it back through SPI. A different design than DVP with only 1 bit data output.

But the Raspberry Pi driver could be made sensor agnostic.

josuah commented:
> I am not sure what is the best solution to resolve the resolution issue with the other dma engines. The HM01B0 documentation states

It could be interesting to declare the size as 324x324 then...

And then in the PIO driver, it will be possible to make use of the `video_set_selection()` API to decide how many pixels to crop the image, effectively using PIO as a DMA 2D engine. \:)

But this could be left as a future PR as well: considering the cropping feature as an improvement, as the current PR did stay in review for long already.

josuah reviewed (dismissed):
Thank you for the reminder. To me it is in very good shape to be merged.
Some existing sensors of Zephyr are not as up to date regarding APIs.

There was a small detail to verify between number of pixels transmitted vs configured, but this seems more like something to keep track of down the line.

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -0,0 +1,301 @@
+		LOG_ERR("Failed to write BIT_CONTROL reg (%d)", ret);
+		return ret;
+	}
+	/* OSC_CLK_DIV */
```
Nitpicking: In other to be consistent and make it easier to read, maybe add a blank line between sections.

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -0,0 +1,301 @@
+
+static int hm01b0_get_caps(const struct device *dev, struct video_caps *caps)
+{
+	caps->min_vbuf_count = 0;
```
Sensors drivers should not set the min_vbuf_count which is something related to memory hence camera receiver rather than sensor.
Moreover min_line_count / max_line_count do not exist anymore so should be removed.

avolmat-st commented on drivers/video/hm01b0.c:
```diff
@@ -0,0 +1,301 @@
+		LOG_ERR("Failed to write GRP_PARAM_HOLD reg (%d)", ret);
+		return ret;
+	}
+	return ret;
```
Nitpicking: global to the whole driver, could you add a blank line before the last return ?

avolmat-st reviewed (changes requested):
Thanks for the patch.
Looks good to me. Few nitpicking comments.
2 points regarding caps (min/max_line_count) / fmt (pitch) to be corrected.

NinoSc commented:
> Thanks for the patch. Looks good to me. Few nitpicking comments. 2 points regarding caps (min/max_line_count) / fmt (pitch) to be corrected.

Hi @avolmat-st, thanks for your review, I addressed your comment and rebased

avolmat-st reviewed (approved):
Few ret != 0 not corrected into ret < 0, however this is minor hence not blocking so apart from that it looks good to me.
